test/jdk/com/sun/jdi/JdwpOnThrowTest.java

```diff
@@ -0,0 +1,161 @@
+            try {
+                vm = attach("localhost", "" + port);
+                EventQueue queue = vm.eventQueue();
+                log("Waiting for exception event");
+                long start = System.currentTimeMillis();
+                while (start + TIMEOUT > System.currentTimeMillis()) {
+                    EventSet eventSet = queue.remove(TIMEOUT);
+                    EventIterator eventIterator = eventSet.eventIterator();
+                    while(eventIterator.hasNext() && start + TIMEOUT > System.currentTimeMillis()) {
+                        Event event = eventIterator.next();
+                        if (event instanceof ExceptionEvent ex) {
+                            if (ex.exception() == null) {
+                                throw new RuntimeException("Exception is null");
+                            }
+                            if (ex.exception().type() == null) {
+                                throw new RuntimeException("Exception type is null");
+                            }
+                            if (ex.exception().referenceType() == null) {
+                                throw new RuntimeException("Exception reference type is null");
+                            }
+                            if (ex.catchLocation() == null) {
+                                throw new RuntimeException("Exception catch location is null");
+                            }
+                            if (!ex.location().equals(ex.thread().frame(0).location())) {
+                                throw new RuntimeException(
+                                    String.format("Throw location %s and location of first frame %s are not equal",
+                                                  ex.location(), ex.thread().frame(0).location()));
+                            }
+                            if (!ex.exception().type().name().equals("Ex")) {
+                                throw new RuntimeException("Exception has wrong type: " + ex.exception().type().name());
+                            }
+                            log("Received exception event: " + event);
+                            vm.dispose();
+                            return;
+                        }
+                        log("Received event: " + event);
+                    }
+                }
+                throw new RuntimeException("ERROR: failed to receive exception event");
+            } catch (IOException ex) {
+                throw new RuntimeException("ERROR: failed to attach", ex);
+            }
+        }
+    }
+
+    private static int findFreePort() {
+        try (ServerSocket socket = new ServerSocket(0)) {
+            return socket.getLocalPort();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static String ATTACH_CONNECTOR = "com.sun.jdi.SocketAttach";
+    // cache socket attaching connector
+    private static AttachingConnector attachingConnector;
```

Comment:
I would have expected a ListeningConnector to be used, so I'm a bit unsure how the attaching is working here and who is listening and who is initiating the connection. I don't understand some of these finer details of connectors and their args that well, but I would think that when using `onthrow` you would tell the debuggee what debugger port it should open the socket on, and the debugger would be listening on that port. Instead it seems you are telling the debuggee what port to listen on and the debugger is opening a socket on that port. Does that mean the open is done by the debugger before the debuggee even throws the exception, and that just leaves the debugger blocked on the open until the debuggee does an accept?

Reply:
> Does that mean the open is done by the debugger before the debuggee even throws the exception, and that just leaves the debugger blocked on the open until the debuggee does an accept?

Yes. This is for three reasons:

1. server mode is the most commonly used mode
2. it is used by the `OnThrowTest` which is the other test that checks that the JDWP agent properly runs the `launch` command after the specified exception has been thrown
3. The original intent of onthrow is, is that the `launch` command launches the debugger (by running it in a new session via screen or tmux), this is why setting the `launch` option is mandatory when using the `onthrow` option. Nowadays most people use onthrow with an IDE, so we could think about removing the requirement.

Reply:
Ok Thanks for the explanation. Maybe another thing that could be documented a bit better.